### DISCUSSION TITLE
How to update entities in many-to-many to match an array.

### DISCUSSION THREAD

**Discussion by computebender** [prisma/prisma#7211] [Q&A] [answered]
Hello!

I am making an app that needs to sync information about features and the versions they belong to from a remote API. These versions are returned as a list corresponding to the version IDs. These versions are able to be added and removed from a feature, so each time a call is made to get the most up to date information the database should reflect that. 

As an example this is what you might get the first time you make a call to the API,
```js
const feature1update1 = {
    featureId: 1,
    featureName: 'Feature 1',
    versions: [1, 2] 
}
```

And later you might get back,
```js
const feature1update2 = {
    featureId: 1,
    featureName: 'Feature 1',
    versions: [1, 2, 3] 
}
```

A third call returns,
```js
const feature1update3 = {
    featureId: 1,
    featureName: 'Feature 1',
    versions: [1] 
}
```

Currently I have a many-to-many relationship between versions and features in my schema that looks like,
```prisma
model Version {
    id           Int        @id @default(autoincrement())
    features     Feature[]
}

model Feature {
    id               Int          @id @default(autoincrement())
    featureName      String
    versions         Version[]
}
```

I am struggling to figure out the best way to keep the versions for a given feature in sync with what the API provides. This is my current best attempt that runs each time a call to the API is made to update the feature,

```js
await prisma.feature.upsert({
    where: {
        id: feature1update1.featureId
    },
    update: {
        featureName: feature1update1.featureName,
        versions: {
            set: feature1update1.versions.map(v => ({id: v}))
        }
    },
    create: {
        featureName: feature1update1.featureName,
        versions: {
            connect: feature1update1.versions.map(v => ({id: v}))
        }
    }
})
```
I know this is wrong because I don't see a way this could handle removing versions from a feature, as well as the fact that it just fails with an error anyways. I'll include the error below, even though I know it doesn't matter because the implementation I have is probably totally wrong,
```
  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Error { kind: Db, cause: Some(DbError { severity: "ERROR", parsed_severity: None, code: SqlState("42601"), message: "syntax error at or near \"ON\"", detail: None, hint: None, position: Some(Original(65)), where_: None, schema: None, table: None, column: None, datatype: None, constraint: None, file: Some("src\\backend\\parser\\scan.l"), line: Some(1054), routine: Some("scanner_yyerror") }) }) })
```

Not only could the versions on a feature change, it could be an entirely new issue which is why I have it in the upsert.

Any guidance or suggestions would be appreciated! I apologize if this is an easy answer, I've just started learning Prisma this week.

**Comment by ryands17** [comment]
@computebender 👋 

I tried the above query and it works fine for me. This will also handle removing versions as `set` will add the ones you specify and remove the ones you don't.

I would suggest running this with sample data already present and making sure `prisma` and `@prisma/client` are on the latest version and in sync i.e. _2.23.0_

**Reply by computebender** [reply]
I did have both packages updated, turns out the issue is when connecting to a Postgres 9.3 server. When I tried the exact same with a Postgres 13 server it works perfectly fine. I was just running 9.3 in my dev env for a different project that specifically requires it, so upgrading this project to Postgres 13 is no problem.

**Reply by ryands17** [reply, marked as answer]
We do [support Postgres 9](https://www.prisma.io/docs/reference/database-reference/supported-databases) so this seems like a bug. It would be great if you could open an issue [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=kind%2Fbug&template=bug_report.yml) so that we can look into this :)
